File: app/src/main/java/com/example/anpu/circles/LogInActivity.java
package com.example.anpu.circles;

import android.animation.Animator;
import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.animation.PropertyValuesHolder;
import android.animation.ValueAnimator;
import android.animation.ValueAnimator.AnimatorUpdateListener;
import android.app.ActivityOptions;
import android.content.Intent;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.transition.Explode;
import android.transition.Fade;
import android.transition.Slide;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.view.animation.AccelerateDecelerateInterpolator;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;

import java.io.IOException;
import java.lang.reflect.Type;
import java.util.HashMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnTextChanged;
import model.User;
import model.UserStatus;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class LogInActivity extends AppCompatActivity {

//    // bind userEditText and pwdEditText
//    @BindView(R.id.edit_username_login) EditText userEditText;
//    @BindView(R.id.edit_pwd_login) EditText pwdEditText;
//    // bind buttons
//    @BindView(R.id.login_button) Button loginButton;
//    // bind textview
//    @BindView(R.id.signup_textview) TextView signupTextView;
//    @BindView(R.id.forgetpwd_textview) TextView forgetpwdTextView;
    @BindView(R.id.main_btn_login) TextView mBtnLogin;
    @BindView(R.id.layout_progress) View progress;
    @BindView(R.id.input_layout) View mInputLayout;
    @BindView(R.id.login_layout_email) LinearLayout mName;
    @BindView(R.id.login_layout_pwd) LinearLayout mPsw;

    private float mWidth, mHeight;

    private ActionBar bar;


    private String email;
    private String pwd;
    private String urlLogin = "http://steins.xin:8001/auth/login";
    private String urlSignup = "http://steins.xin:8001/auth/signup";
    private String urlForget = "http://steins.xin:8001/auth/forgetpwd";



    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        // new
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);  // bind userEditText and pwdEditText

        // hide the bar at the top
//        getSupportActionBar().hide();
        bar = getSupportActionBar();
        if (bar != null) {
            bar.hide();
        }

        Fade fade = new Fade();
        fade.setDuration(1000);

        Explode explode = new Explode();
        explode.setDuration(1000);

        Slide slide = new Slide();
        slide.setDuration(1000);

//        getWindow().setEnterTransition(fade);
        getWindow().setExitTransition(slide);
    }

    @OnTextChanged(R.id.edit_email_login)
    void nameChanged(CharSequence s, int start, int before, int count) {
        email = s.toString();
    }

    @OnTextChanged(R.id.edit_pwd_login)
    void pwdChanged(CharSequence s, int start, int before, int count) {
        pwd = s.toString();
    }

    @OnClick(R.id.main_btn_login)
    void loginClicked() {

        if (email == null || email.equals("")) {
            Toast.makeText(this, "Email is empty", Toast.LENGTH_SHORT).show();
        }
        else if (pwd == null || pwd.equals("")) {
            Toast.makeText(this, "Password is empty", Toast.LENGTH_SHORT).show();
        }
        else {
            // Using regex to check if it's nyu email
            checkEmail();
        }


    }


    @OnClick(R.id.signup_textview)
    void signupTextClicked() {
        Intent intent = new Intent(LogInActivity.this, SignUpActivity.class);
        intent.putExtra("transition", "slide");
        startActivity(intent, ActivityOptions.makeSceneTransitionAnimation(this).toBundle());
    }


    @OnClick(R.id.forgetpwd_textview)
    void forgetpwdTextClicked() {
        if (email != null && !email.isEmpty()) {
            final Gson gsonForget = new GsonBuilder().enableComplexMapKeySerialization().create();
            HashMap<String, String> forget = new HashMap<>();
            forget.put("email", email);


            OkHttpClient client = new OkHttpClient();
            RequestBody body = RequestBody.create(MediaType.parse("application/json; charset=utf-8")
                    , gsonForget.toJson(forget));
            final Request request = new Request.Builder()
                    .post(body)
                    .url(urlForget)
                    .build();
            client.newCall(request).enqueue(new Callback() {
                @Override
                public void onFailure(Call call, IOException e) {
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            Toast.makeText(LogInActivity.this, "Failure to connect to the server", Toast.LENGTH_LONG).show();
                        }
                    });
                }

                @Override
                public void onResponse(Call call, Response response) throws IOException {
                    final Gson resp = new GsonBuilder().enableComplexMapKeySerialization().create();
                    Type type = new TypeToken<HashMap<String, String>>() {}.getType();
                    HashMap<String, String> resHash = resp.fromJson(response.body().string(), type);
                    if (resHash.get("status").equals("1")) {
                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                Toast.makeText(LogInActivity.this, "An email has sent to reset the password.", Toast.LENGTH_LONG).show();
                            }
                        });
                    } else {
                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                Toast.makeText(LogInActivity.this, "Please check your email it doesn't seem to be right.", Toast.LENGTH_LONG).show();
                            }
                        });
                    }
                }
            });
        }
    }

    private void checkEmail() {
        String pattern = "@nyu.edu";
        Pattern r = Pattern.compile(pattern);
        Matcher m = r.matcher(email);
        if (! m.find()) {
            recovery();
            Toast.makeText(this, "Email should end with @nyu.edu", Toast.LENGTH_LONG).show();
        }
        else {
            mWidth = mBtnLogin.getMeasuredWidth();
            mHeight = mBtnLogin.getMeasuredHeight();

            mName.setVisibility(View.INVISIBLE);
            mPsw.setVisibility(View.INVISIBLE);

            inputAnimator(mInputLayout, mWidth, mHeight);
        }
    }

    private void inputAnimator(final View view, float w, float h) {

        AnimatorSet set = new AnimatorSet();

        ValueAnimator animator = ValueAnimator.ofFloat(0, w);
        animator.addUpdateListener(new AnimatorUpdateListener() {

            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                float value = (Float) animation.getAnimatedValue();
                ViewGroup.MarginLayoutParams params = (ViewGroup.MarginLayoutParams) view.getLayoutParams();
                params.leftMargin = (int) value;
                params.rightMargin = (int) value;
                view.setLayoutParams(params);
            }
        });

        ObjectAnimator animator2 = ObjectAnimator.ofFloat(mInputLayout, "scaleX", 1f, 0.5f);
        set.setDuration(500);
        set.setInterpolator(new AccelerateDecelerateInterpolator());
        set.playTogether(animator, animator2);
        set.start();
        set.addListener(new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animator) {

            }

            @Override
            public void onAnimationEnd(Animator animator) {

                progress.setVisibility(View.VISIBLE);
                progressAnimator(progress);
                mInputLayout.setVisibility(View.INVISIBLE);

            }

            @Override
            public void onAnimationCancel(Animator animator) {

            }

            @Override
            public void onAnimationRepeat(Animator animator) {

            }
        });
    }

    private void progressAnimator(final View view) {
        PropertyValuesHolder animator = PropertyValuesHolder.ofFloat("scaleX", 0.5f, 1f);
        PropertyValuesHolder animator2 = PropertyValuesHolder.ofFloat("scaleY", 0.5f, 1f);
        ObjectAnimator animator3 = ObjectAnimator.ofPropertyValuesHolder(view,
                animator, animator2);
        animator3.setDuration(1000);
        animator3.setInterpolator(new JellyInterpolator());
//        animator3.setInterpolator(new AccelerateDecelerateInterpolator());
        animator3.addListener(new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animator) {

            }

            @Override
            public void onAnimationEnd(Animator animator) {
                // check if the account is valid
                // generate json
                validLogin();

            }

            @Override
            public void onAnimationCancel(Animator animator) {

            }

            @Override
            public void onAnimationRepeat(Animator animator) {

            }
        });
        animator3.start();
    }

    private void recovery() {
        progress.setVisibility(View.GONE);
        mInputLayout.setVisibility(View.VISIBLE);
        mName.setVisibility(View.VISIBLE);
        mPsw.setVisibility(View.VISIBLE);

        ViewGroup.MarginLayoutParams params = (ViewGroup.MarginLayoutParams) mInputLayout.getLayoutParams();
        params.leftMargin = 0;
        params.rightMargin = 0;
        mInputLayout.setLayoutParams(params);

        ObjectAnimator animator2 = ObjectAnimator.ofFloat(mInputLayout, "scaleX", 0.5f, 1f);
        animator2.setDuration(500);
        animator2.setInterpolator(new AccelerateDecelerateInterpolator());
        animator2.start();
    }

    private void validLogin() {
        final Gson gson = new Gson();
        User user = new User(email, pwd);
        String jsonUser = gson.toJson(user);
        // post to the server
        OkHttpClient client = new OkHttpClient();
        RequestBody body = RequestBody.create(MediaType.parse("application/json; charset=utf-8"), jsonUser);
        Request request = new Request.Builder()
                .post(body)
                .url(urlLogin)
                .build();
        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        Toast.makeText(LogInActivity.this, "Failure to connect to the server", Toast.LENGTH_LONG).show();
                        recovery();
                    }
                });
            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {
                UserStatus userStatus = gson.fromJson(response.body().string(), UserStatus.class);
                // failure
                if (userStatus.getStatus() == 0) {
                    if (userStatus.getType() == 0) {
                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                Toast.makeText(LogInActivity.this, "Account is not activated.", Toast.LENGTH_LONG).show();
                                recovery();
                            }
                        });
                    }
                    else {
                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                Toast.makeText(LogInActivity.this, "Account does not exist.", Toast.LENGTH_LONG).show();
                                recovery();
                            }
                        });
                    }
                }
                // success
                else {
                    Intent intent = new Intent(LogInActivity.this, HomePage1.class);
                    startActivity(intent);
                }
            }
        });
    }






    /// three listeners for text change   ---------- IGNORE IT
//    @OnTextChanged(value = R.id.edit_username)
//    void onTextChanged(CharSequence s, int start, int before, int count) {
//        Log.d("Test", s.toString());
//    }

//    void afterTextChanged(Editable e) {
//        Toast.makeText(this, e.toString(), Toast.LENGTH_SHORT).show();
//    }
//    void beforeTextChanged(CharSequence s, int start, int count, int after) {
//        Log.d("Test", s.toString());
//    }

//    @OnTextChanged(value = R.id.edit_pwd, callback = OnTextChanged.Callback.TEXT_CHANGED)
//    public void  getPwd(EditText pwd) {
//        Log.d("Test", String.valueOf(pwd.getText()));
//    }


//    @OnTextChanged(R.id.edit_username_login)
//    void usernameChanged(CharSequence s, int start, int before, int count) {
//        email = s.toString();
//    }
//
//    @OnTextChanged(R.id.edit_pwd_login)
//    void pwdChanged(CharSequence s, int start, int before, int count) {
//        pwd = s.toString();
//    }
//
//    @OnClick(R.id.login_button)
//    void loginClicked() {
//        if (email == null || email.equals("")) {
//            Toast.makeText(this, "Username is empty", Toast.LENGTH_SHORT).show();
//        }
//        else if (pwd == null || pwd.equals("")) {
//            Toast.makeText(this,"Password is empty", Toast.LENGTH_SHORT).show();
//        }
//        else {
//            // TO DO
//            // check if email is valid
//            String pattern = "@nyu.edu";
//            Pattern r = Pattern.compile(pattern);
//            Matcher m = r.matcher(email);
//            if (! m.find()) {
//                Toast.makeText(this, "Username should end with @nyu.edu", Toast.LENGTH_SHORT).show();
//            }
//        }
//    }
//

}

File: app/src/main/java/com/example/anpu/circles/LoginActivity.kt
package com.example.anpu.circles

import android.animation.*
import android.app.ActivityOptions
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.preference.PreferenceManager
import android.support.v7.app.AppCompatActivity
import android.text.Editable
import android.text.TextWatcher
import android.transition.Slide
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.view.WindowManager
import android.view.animation.AccelerateDecelerateInterpolator
import android.widget.EditText
import com.bumptech.glide.Glide
import com.example.anpu.circles.model.User
import com.example.anpu.circles.model.UserData
import com.example.anpu.circles.model.UserResponseStatus
import com.example.anpu.circles.utilities.JellyInterpolator
import com.example.anpu.circles.utilities.MD5Util
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import kotlinx.android.synthetic.main.activity_login.*
import kotlinx.android.synthetic.main.input_layout_login.*
import okhttp3.*
import org.jetbrains.anko.sdk25.coroutines.onClick
import org.jetbrains.anko.toast
import java.io.IOException
import java.util.*
import java.util.regex.Pattern

class LoginActivity : AppCompatActivity() {

    private var email : String? = ""
    private var pwd : String? = ""
    private val urlLogin = "http://steins.xin:8001/auth/login"
    private val urlForget = "http://steins.xin:8001/auth/forgetpwd"
    private var sprefLogin : SharedPreferences? = null
    private var editorLogin : SharedPreferences.Editor? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        window.setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN)
        setContentView(R.layout.activity_login)


        val bar = supportActionBar
        bar?.hide()

        val slide = Slide()
        slide.duration = 1000
        window.exitTransition = slide

        sprefLogin = PreferenceManager.getDefaultSharedPreferences(this)
        editorLogin = sprefLogin?.edit()
        editorLogin?.apply()

        Glide.with(applicationContext).load(R.drawable.paw_code).into(login_avatar)

        //listener
        edit_email_login.setTextChangeListener {
            email = it
        }

        edit_pwd_login.setTextChangeListener {
            pwd = it
        }

        main_btn_login.onClick {
            val emailLength = email?.length
            val pwdLength = pwd?.length

            if (emailLength == null || emailLength == 0) {
                toast("Email is empty")
            } else if (pwdLength == null || pwdLength == 0) {
                toast("Password is empty")
            } else {
                checkEmail()
            }
        }

        signup_textview.onClick {
            val intent = Intent(this@LoginActivity, SignUpActivity::class.java)
            intent.putExtra("transition", "slide")
            startActivity(intent, ActivityOptions.makeSceneTransitionAnimation(this@LoginActivity).toBundle())
        }

        forgetpwd_textview.onClick {
            val emailLength = email?.length
            if (emailLength != null && emailLength > 0) {
                val gsonForget = GsonBuilder().enableComplexMapKeySerialization().create()
                val forget = HashMap<String, String>()
                forget.put("email", email!!)
                val client = OkHttpClient()
                val body = RequestBody.create(MediaType.parse("application/json; charset=utf-8"), gsonForget.toJson(forget))
                val request = Request.Builder()
                        .post(body)
                        .url(urlForget)
                        .build()
                client.newCall(request).enqueue(object : Callback {
                    override fun onFailure(call: Call, e: IOException) {
                        runOnUiThread { toast("Failure to connect to the server") }
                    }
                    @Throws(IOException::class)
                    override fun onResponse(call: Call, response: Response) {
                        val resp = GsonBuilder().enableComplexMapKeySerialization().create()
                        val type = object : TypeToken<HashMap<String, String>>() {}.type
                        val resHash = resp.fromJson<HashMap<String, String>>(response.body().string(), type)
                        if (resHash["status"] == "1") {
                            runOnUiThread { toast("An email has sent to reset the password.")}
                        } else {
                            runOnUiThread { toast("Please check your email it doesn't seem to be right.")}
                        }
                    }
                })

            }
        }


    }

    private fun checkEmail() {
        val pattern = "@nyu.edu"
        val r = Pattern.compile(pattern)
        val m = r.matcher(email)
        if (!m.find()) {
            toast("Email should end with @nyu.edu")
        } else {
            val mWidth = main_btn_login.measuredWidth.toFloat()
            val mHeight = main_btn_login.measuredHeight.toFloat()

            login_layout_email.visibility = View.INVISIBLE
            login_layout_pwd.visibility = View.INVISIBLE

            inputAnimator(input_layout_login, mWidth, mHeight)
        }
    }

    private fun inputAnimator(view: View, w: Float, h: Float) {

        val set = AnimatorSet()

        val animator = ValueAnimator.ofFloat(0f, w)
        animator.addUpdateListener { animation ->
            val value = animation.animatedValue as Float
            val params = view.layoutParams as ViewGroup.MarginLayoutParams
            params.leftMargin = value.toInt()
            params.rightMargin = value.toInt()
            view.layoutParams = params
        }

        val animator2 = ObjectAnimator.ofFloat(input_layout_login, "scaleX", 1f, 0.5f)
        set.duration = 500
        set.interpolator = AccelerateDecelerateInterpolator()
        set.playTogether(animator, animator2)
        set.start()
        set.addListener(object : Animator.AnimatorListener {
            override fun onAnimationStart(animator: Animator) {

            }

            override fun onAnimationEnd(animator: Animator) {

                layout_progress_login.visibility = View.VISIBLE
                progressAnimator(layout_progress_login)
                input_layout_login.visibility = View.INVISIBLE

            }

            override fun onAnimationCancel(animator: Animator) {

            }

            override fun onAnimationRepeat(animator: Animator) {

            }
        })
    }

    private fun recovery() {
        layout_progress_login.visibility = View.GONE
        input_layout_login.visibility = View.VISIBLE
        login_layout_email.visibility = View.VISIBLE
        login_layout_pwd.visibility = View.VISIBLE

        val params = input_layout_login.layoutParams as ViewGroup.MarginLayoutParams
        params.leftMargin = 0
        params.rightMargin = 0
        input_layout_login.layoutParams = params

        val animator2 = ObjectAnimator.ofFloat(input_layout_login, "scaleX", 0.5f, 1f)
        animator2.duration = 500
        animator2.interpolator = AccelerateDecelerateInterpolator()
        animator2.start()
    }

    private fun progressAnimator(view: View) {
        val animator = PropertyValuesHolder.ofFloat("scaleX", 0.5f, 1f)
        val animator2 = PropertyValuesHolder.ofFloat("scaleY", 0.5f, 1f)
        val animator3 = ObjectAnimator.ofPropertyValuesHolder(view,
                animator, animator2)
        animator3.duration = 1000
        animator3.interpolator = JellyInterpolator()
        animator3.addListener(object : Animator.AnimatorListener {
            override fun onAnimationStart(animator: Animator) {

            }

            override fun onAnimationEnd(animator: Animator) {
                validLogin()
            }

            override fun onAnimationCancel(animator: Animator) {

            }

            override fun onAnimationRepeat(animator: Animator) {

            }
        })
        animator3.start()
    }

    private fun validLogin() {
        val gson = Gson()
        val user = User(email, MD5Util.getMD5Str(pwd))
        val jsonUser = gson.toJson(user)
        // post to the server
        val client = OkHttpClient()
        val body = RequestBody.create(MediaType.parse("application/json; charset=utf-8"), jsonUser)
        val request = Request.Builder()
                .post(body)
                .url(urlLogin)
                .build()
        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                runOnUiThread {
                    toast("Failure to connect to the server")
                    recovery()
                }
            }

            @Throws(IOException::class)
            override fun onResponse(call: Call, response: Response) {

                val userResponseStatus = gson.fromJson(response.body().string(), UserResponseStatus::class.java)
                // failure
                if (userResponseStatus.status == 0) {
                    if (userResponseStatus.type == 0) {
                        runOnUiThread {
                            toast("Account is not activated.")
                            recovery()
                        }
                    } else {
                        runOnUiThread {
                            toast("Account does not exist.")
                            recovery()
                        }
                    }
                } else {
                    UserData.setEmail(MD5Util.getMD5(email))
                    UserData.setNickname(userResponseStatus.nickname)
                    UserData.setAvatar(userResponseStatus.avatar)
                    UserData.setUncypheredEmail(email)

                    editorLogin?.putString("email", email)
                    editorLogin?.commit()

                    val intent = Intent(this@LoginActivity, HomePageFragmentActivity::class.java)
                    startActivity(intent)
                    this@LoginActivity.finish()
                }
            }
        })
    }


    private fun EditText.setTextChangeListener(body: (key: String) -> Unit) {
        addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(s: Editable?) {
            }

            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                body(s.toString())
            }
        })
    }
}